c_glib/arrow-glib/record-batch.cpp

```diff
@@ -402,6 +408,173 @@ garrow_record_batch_remove_column(GArrowRecordBatch *record_batch,
   }
 }
 
+
+typedef struct GArrowRecordBatchIteratorPrivate_ {
+  arrow::RecordBatchIterator iterator;
+} GArrowRecordBatchIteratorPrivate;
+
+enum {
+  PROP_ITERATOR = 1,
+};
+
+G_DEFINE_TYPE_WITH_PRIVATE(GArrowRecordBatchIterator,
+                           garrow_record_batch_iterator,
+                           G_TYPE_OBJECT)
+
+#define GARROW_RECORD_BATCH_ITERATOR_GET_PRIVATE(obj)        \
+  static_cast<GArrowRecordBatchIteratorPrivate *>(           \
+     garrow_record_batch_iterator_get_instance_private(      \
+       GARROW_RECORD_BATCH_ITERATOR(obj)))
+
+static void
+garrow_record_batch_iterator_finalize(GObject *object)
+{
+  auto priv = GARROW_RECORD_BATCH_ITERATOR_GET_PRIVATE(object);
+
+  priv->iterator.~Iterator();
+
+  G_OBJECT_CLASS(garrow_record_batch_iterator_parent_class)->finalize(object);
+}
+
+static void
+garrow_record_batch_iterator_set_property(GObject *object,
+                                          guint prop_id,
+                                          const GValue *value,
+                                          GParamSpec *pspec)
+{
+  auto priv = GARROW_RECORD_BATCH_ITERATOR_GET_PRIVATE(object);
+
+  switch (prop_id) {
+  case PROP_ITERATOR:
+    priv->iterator =
+      std::move(*static_cast<arrow::RecordBatchIterator *>(g_value_get_pointer(value)));
+    break;
+  default:
+    G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+    break;
+  }
+}
+
+static void
+garrow_record_batch_iterator_init(GArrowRecordBatchIterator *object)
+{
+  auto priv = GARROW_RECORD_BATCH_ITERATOR_GET_PRIVATE(object);
+  new(&priv->iterator) arrow::RecordBatchIterator;
+}
+
+static void
+garrow_record_batch_iterator_class_init(GArrowRecordBatchIteratorClass *klass)
+{
+  auto gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->finalize     = garrow_record_batch_iterator_finalize;
+  gobject_class->set_property = garrow_record_batch_iterator_set_property;
+
+  GParamSpec *spec;
+
+  spec = g_param_spec_pointer("iterator",
+                              "Iterator",
+                              "The raw arrow::RecordBatchIterator",
+                              static_cast<GParamFlags>(G_PARAM_WRITABLE |
+                                                       G_PARAM_CONSTRUCT_ONLY));
+  g_object_class_install_property(gobject_class, PROP_ITERATOR, spec);
+}
+
+/**
+ * garrow_record_batch_iterator_new:
+ * @record_batches: (element-type GArrowRecordBatch):
+ *   The record batches.
+ *
+ * Returns: A newly created #GArrowRecordBatchIterator.
+ *
+ * Since: 0.17.0
+ */
+GArrowRecordBatchIterator *
+garrow_record_batch_iterator_new(GList *record_batches)
+{
+  std::vector<std::shared_ptr<arrow::RecordBatch>> arrow_record_batches;
```

Comment:
I came up with the idea that it can avoid making a vector to introduce `GListIterator<T>` like `arrow::VectorIterator<T>`.

@kou How do you think about this?

Reply:
It's a good idea!

Reply:
FYI: I've removed temporary vector in `garrow_record_batch_iterator_to_list()`.

Reply:
> I came up with the idea that it can avoid making a vector to introduce `GListIterator<T>` like `arrow::VectorIterator<T>`.
> 
> @kou How do you think about this?

How do you create a Glist  of anything, especially fields. It is not clear how do just looking at the api funtions available. Please I would be gratefull. 

What I  understand so far is, 

To create a GList of fields,  your need this funtionn call -> 
GList * garrow_struct_data_type_get_fields (GArrowStructDataType *struct_data_type);

which means you have to create a GArrowStructDatatype -> which is dont with this function call

GArrowStructDataType *
garrow_struct_data_type_new (GList *fields);, which also depends on GList, I am a bit confused, how to a funtion depends on another function that depends on it self.

Reply:
See https://github.com/apache/arrow/blob/master/c_glib/arrow-glib/schema.cpp#L241-L247, this is a typical pattern to make a `GList` from scratch.

By the way, please do not ask unrelated things to this pull-request here.
You can ask anything about Apache Arrow in the mailing list.
http://mail-archives.apache.org/mod_mbox/arrow-dev/

Reply:
thank you very much, I am sorry  asking here.

Reply:
@kou I noticed `GListIterator` is not usable for implementing `GArrowRecordBatchIterator` because it is a binding of single `arrow::RecordBatchIterator` instead of a list of `GArrowRecordBatch`.

I postponed to introduce `GListIterator` until we actually need it.

Reply:
OK. Then I'll merge this pull request.

Reply:
@kou I noticed again that `GListIterator` can be used just only for `garrow_record_batch_iterator_new`.  I was confused.

Reply:
OK. We can work on it as a follow-up pull request.

Reply:
OK!